src/main/java/dev/mccue/resolve/graph/DependencyTree.java

```diff
@@ -1,20 +1,101 @@
 
-    boolean excluded();
+    //TODO roots0 can be optionally null
+    static List<DependencyTree> of(
+            Resolution resolution,
+            List<Dependency> roots,
+            boolean withExclusions
+    ) {
+        //TODO Delete after you write code
+        record Person(String name, int age) {
+        }
 
-    String reconciledVersion();
-    String retainedVersion();
+        List<Person> persons = new ArrayList<>();
+        persons.sort(Comparator.comparing(Person::age));
+        persons.sort(Comparator.nullsFirst(Comparator.comparing(Person::name)).thenComparing(Comparator.comparing(Person::age)));
 
-    List<DependencyTree> children();
+        Set<Integer> ages = Set.of();
+
+        persons.stream()
+                .sorted(Comparator.comparing(Person::age))
+                .filter(person -> !ages.contains(person.age))
+                .map(person -> new Object[]{person.age})
+                .toList();
+
+
+        List<Dependency> roots0;
+        if (!roots.isEmpty()) {
+            roots0 = resolution.rootDependencies;
+        } else {
+            roots0 = roots;
+        }
+
+        return roots0.stream().<DependencyTree>map(dep -> new Node(dep, false, resolution, withExclusions)).toList();
+    }
+
+    record Node(
+            Dependency dependency,
+            Boolean excluded,
+            Resolution resolution,
+            Boolean withExclusions
+    ) implements DependencyTree {
+        String reconciledVersion() {
+            if (resolution.reconciledVersion.dependency.module() != null) {
+                return resolution.reconciledVersion.dependency.module().toString();
+            } else {
+                return resolution.reconciledVersion.dependency.version();
+            }
+        }
+
+        String retainedVersion() {
+            if (resolution.retainedVersion.dependency.module() != null) {
+                return resolution.retainedVersion.dependency.module().toString();
+            } else {
+                return resolution.retainedVersion.dependency.version();
+            }
+        }
+
+        List<DependencyTree> children() {
+            if (excluded) {
+                return null;
+            } else {
+                var dep0 = dependency.withVersion(retainedVersion());
+
+                List<Dependency> dependencies = resolution.dependenciesOf(dep0, false)
+                        .stream()
+                        .sorted(Comparator.comparing((Dependency trDep) ->
+                                trDep.module().organization()).thenComparing(trDep ->
+                                trDep.module().name()).thenComparing(trDep -> trDep.version()))
+                        .collect(Collectors.toList());
+
+                List dependencies0 = dependencies.stream().map(Dependency::moduleVersion).collect(Collectors.toList());
+
+                var excluded = resolution.dependenciesOf(dep0.withExclusions(Collections.emptyList(), false))
```

Comment:
You can use `List.of()` instead of `Collections.emptyList()` if you want

Reply:
Will change